File: clients/apps/lib/services/conversation_service.dart
import '../models/models.dart';
import '../models/requests.dart';
import '../network/conversation_api_client.dart';

class ConversationService {
  final ConversationApiClient _client;

  ConversationService(this._client);

  Future<List<Conversation>> getConversations() async {
    final res = await _client.getConversations();
    return res.data.data?.conversations ?? [];
  }

  Future<Conversation> createDirectConversation(String recipientId) async {
    final res = await _client.createDirectConversation(
      CreateDirectConversationRequest(recipientId: recipientId),
    );
    return res.data.data!;
  }

  Future<Conversation> createGroupConversation(
    String name,
    List<String> participantIds,
  ) async {
    final res = await _client.createGroupConversation(
      CreateGroupConversationRequest(
        name: name,
        participantIds: participantIds,
      ),
    );
    return res.data.data!;
  }

  Future<void> markAsRead(String conversationId) async {
    await _client.markAsRead(conversationId);
  }

  Future<Conversation?> checkDirectConversation(String recipientId) async {
    try {
      final res = await _client.checkDirectConversation(recipientId);
      return res.data.data;
    } catch (_) {
      return null;
    }
  }

  Future<Conversation> getConversationDetail(String conversationId) async {
    final res = await _client.getConversationDetail(conversationId);
    return res.data.data!;
  }

  Future<void> sendTyping(String conversationId) async {
    await _client.sendTyping(TypingRequest(conversationId: conversationId));
  }
}

File: clients/packages/shared/src/services/conversation.service.ts
import type {
  ApiResponse,
  Conversation,
  ConversationDetail,
  ConversationsListResponse,
  CreateGroupConversationDTO,
} from '../types'
import { http } from './http'

export const conversationService = {
  async getConversations(): Promise<ApiResponse<ConversationsListResponse>> {
    const res = await http.get<ApiResponse<ConversationsListResponse>>('/conversations')
    return res.data
  },

  async getConversation(id: string): Promise<ApiResponse<ConversationDetail>> {
    const res = await http.get<ApiResponse<ConversationDetail>>(`/conversations/${id}`)
    return res.data
  },

  async createGroupConversation(
    data: CreateGroupConversationDTO,
  ): Promise<ApiResponse<Conversation>> {
    const res = await http.post<ApiResponse<Conversation>>('/conversations/group', data)
    return res.data
  },

  async checkDirectConversation(recipientId: string): Promise<ApiResponse<Conversation>> {
    const res = await http.get<ApiResponse<Conversation>>('/conversations/direct/check', {
      params: { recipientId },
    })
    return res.data
  },

  async createDirectConversation(recipientId: string): Promise<ApiResponse<Conversation>> {
    const res = await http.post<ApiResponse<Conversation>>('/conversations/direct', {
      recipientId,
    })
    return res.data
  },

  async sendTypingIndicator(conversationId: string): Promise<ApiResponse<void>> {
    const res = await http.post<ApiResponse<void>>('/conversations/typing', {
      conversationId,
    })
    return res.data
  },

  async markAsRead(conversationId: string): Promise<ApiResponse<void>> {
    const res = await http.put<ApiResponse<void>>(`/conversations/${conversationId}/read`)
    return res.data
  },

  async hide(conversationId: string): Promise<ApiResponse<void>> {
    const res = await http.post<ApiResponse<void>>(`/conversations/${conversationId}/hide`)
    return res.data
  },

  async unhide(conversationId: string): Promise<ApiResponse<void>> {
    const res = await http.post<ApiResponse<void>>(`/conversations/${conversationId}/unhide`)
    return res.data
  },
}
